packages/deployer/src/utils/compiler.ts

```diff
@@ -1,12 +1,100 @@
+ */
+export function constructUniqueSourceFileId(directoryNamespace: string, sourceFilePath: string): string {
+    const namespacePrefix = !_.isEmpty(directoryNamespace) ? `/${directoryNamespace}` : '';
+    const sourceFilePathNoLeadingSlash = sourceFilePath.replace(/^\/+/g, '');
+    const sourceFileId = `${namespacePrefix}/${sourceFilePathNoLeadingSlash}`;
+    return sourceFileId;
+}
+/**
+ * Constructs a system-wide unique identifier for a dependency file.
+ * @param dependencyFilePath Path from a sourceFile to a dependency.
+ * @param  contractBaseDir Base contracts directory of search tree.
+ * @return sourceFileId A system-wide unique identifier for the source file.
+ */
+export function constructDependencyFileId(dependencyFilePath: string, sourceFilePath: string): string {
+    if (_.startsWith(dependencyFilePath, '/')) {
+        // Path of the form /namespace/path/to/dependency.sol
+        return dependencyFilePath;
+    } else {
+        // Dependency is relative to the source file: ./dependency.sol, ../../some/path/dependency.sol, etc.
+        // Join the two paths to construct a valid source file id: /namespace/path/to/dependency.sol
+        return path.join(path.dirname(sourceFilePath), dependencyFilePath);
+    }
+}
+/**
+ * Constructs a system-wide unique identifier for a contract.
+ * @param directoryNamespace Namespace of the source file's root contract directory.
+ * @param sourceFilePath Path to a source file, relative to contractBaseDir.
+ * @return sourceFileId A system-wide unique identifier for contract.
+ */
+export function constructContractId(directoryNamespace: string, sourceFilePath: string): string {
+    const namespacePrefix = !_.isEmpty(directoryNamespace) ? `${directoryNamespace}:` : '';
+    const sourceFileName = path.basename(sourceFilePath, constants.SOLIDITY_FILE_EXTENSION);
+    const contractId = `${namespacePrefix}${sourceFileName}`;
+    return contractId;
+}
+/**
+ * Overloaded function names are incremented as follows: functionName, functionName_2, functioname_3, ...
+ * If functionName_N already exists then compilation will fail.
+ * @param contractAbi Contract ABI
+ * @return Contract ABI with overloaded functions renamed.
+ */
+export function renameOverloadedFunctions(contractAbi: ContractAbi): ContractAbi {
+    // Collect list of function names in the contract ABI
+    const functionNameList = _.map(contractAbi, abiItem => {
+        const type = abiItem.type;
+        if (type !== AbiType.Function) {
+            return; // skips this abiItem
+        }
+
+        const functionName = (abiItem as MethodAbi).name;
+        return functionName;
+    });
+
+    // Rename overloaded functions
+    const outputContractAbi = _.cloneDeep(contractAbi);
+    const functionNameToSeenCount: FunctionNameToSeenCount = {};
+    _.forEach(outputContractAbi, abiItem => {
+        const type = abiItem.type;
+        if (type !== AbiType.Function) {
```

Comment:
instead of duplicating this check here and line 61, we can just create a filtered list of `MethodAbi` instances using `_.filter` and transform that list into a list of functionNames and then perform the logic here. That will also get rid of the need to cast `abiItem` as `MethodAbi` as well

Reply:
we can also return w/o performing a clone if we see that the list is empty after filtering